File: rabbitmq-orders/producer-python/produce.py
"""Producer (Python) — publishes canonical BabelQueue envelopes to a RabbitMQ
queue that a consumer in *any other language* can read.

    pip install "babelqueue[amqp]"
    python produce.py

Each message is projected onto native AMQP 0-9-1 properties (§2 of the broker-bindings
contract): ``type`` = URN, ``correlation_id`` = trace_id, ``message_id`` = meta.id,
``app_id`` = babelqueue, plus ``x-schema-version`` / ``x-source-lang`` / ``x-attempts``
headers — so the Node consumer routes on ``properties.type`` without parsing the body.

Defaults target the local RabbitMQ from docker-compose. Point BROKER_URL at any
RabbitMQ to run it there.
"""

import os

from babelqueue import BabelQueue

BROKER_URL = os.environ.get("BROKER_URL", "amqp://guest:guest@localhost:5672/")

app = BabelQueue(BROKER_URL, queue="orders")

messages = [
    ("urn:babel:orders:created", {"order_id": 1001, "amount": 19.99}),
    ("urn:babel:orders:created", {"order_id": 1002, "amount": 39.98}),
    ("urn:babel:orders:created", {"order_id": 1003, "amount": 59.97}),
    ("urn:babel:orders:shipped", {"order_id": 1002, "carrier": "DHL"}),
]

for urn, data in messages:
    message_id = app.publish(urn, data)
    print(f"[python] published {urn}  id={message_id}  data={data}")

print(f"[python] {len(messages)} message(s) on the 'orders' RabbitMQ queue — now run the Node consumer.")

File: pulsar-orders/producer-java/src/main/java/Produce.java
import com.babelqueue.pulsar.PulsarPublisher;
import java.util.Map;
import org.apache.pulsar.client.api.Producer;
import org.apache.pulsar.client.api.PulsarClient;

/**
 * Java produces canonical BabelQueue envelopes onto the Pulsar "orders" topic. A .NET
 * service (consumer-dotnet) reads the same topic and routes by URN — same wire envelope,
 * different languages, one broker. The producer sets the §5 bq- properties so the consumer
 * routes on bq-job without decoding the body.
 */
public final class Produce {
    public static void main(String[] args) throws Exception {
        String url = System.getenv().getOrDefault("PULSAR_URL", "pulsar://localhost:6650");

        try (PulsarClient client = PulsarClient.builder().serviceUrl(url).build();
             Producer<byte[]> producer = client.newProducer().topic("orders").create()) {

            PulsarPublisher publisher = PulsarPublisher.create(producer);

            for (int i = 1; i <= 3; i++) {
                int orderId = 1000 + i;
                String id = publisher.publish(
                    "urn:babel:orders:created",
                    Map.of("order_id", orderId, "amount", 19.99 * i));
                System.out.printf("[java] published orders:created order_id=%d meta.id=%s%n", orderId, id);
            }

            String shipId = publisher.publish(
                "urn:babel:orders:shipped",
                Map.of("order_id", 1002, "carrier", "DHL"));
            System.out.printf("[java] published orders:shipped order_id=1002 meta.id=%s%n", shipId);

            System.out.println("[java] done — 4 envelopes on topic 'orders'.");
        }
    }
}
